### DISCUSSION TITLE
4/24 (월) 조이 - 계층화 아키텍처(Layered Architecture)

### DISCUSSION THREAD

**Discussion by yeonkkk** [woowacourse-study/2023-presentation-study#8] [Q&A] [answered]
Q & A

**Comment by sh111-coder** [comment, marked as answer]
Q : 도메인 레이어 정의가 애매합니다!!

Q : 예시에서는 Repository가 없는데 Repository를 만들어야 하는 이유가 있을까요?

Q : 모놀리식 애플리케이션을 자세히 알려주세요

**Reply by yeonkkk** [reply]
안녕하세요 성하! 조이입니다.
질문을 3개나 주셨네요! 하나씩 답변드려보겠습니다.

**Reply by yeonkkk** [reply]
> Q1 : 도메인 레이어 정의가 애매합니다!!

스터디 때 설명드린 것과 같이, 도메인 레이어는 " 도메인의 핵심 개념과 비즈니스 규칙을 나타내는 도메인 모델이 포함되어 있고, 알고리즘 및 프로그래밍 구성 요소를 담당하는 레이어"입니다. 
이런 정의가 궁금하셨던 건 아닌 것 같지만, 남겨주신 질문만으로는 정확히 어떤 점이 궁금한지 모르겠어서 참고하실 수 있을만한 자료를 남기겠습니다!
만약 추가적으로 궁금하신 점이 있다면 답글 주세요!🙂

[Layered Architecture](http://uniknow.github.io/AgileDev/site/0.1.8-SNAPSHOT/parent/ddd/core/layered_architecture.html)
[What is a Layered Architecture in software design?](https://medium.com/geekculture/what-is-a-layered-architecture-in-software-design-22152fc06822)

**Reply by yeonkkk** [reply]
> Q2 : 예시에서는 Repository가 없는데 Repository를 만들어야 하는 이유가 있을까요?

작성한 자료의 예시에서는 `Repository`가 없고, `dao`만 사용하고 있습니다. 질문의 요지가 `repository`와 `dao` 의 차이점을 물으시는 것 같아서, 이에 초점을 맞춰 답변드려보겠습니다. 다만, 해당 내용은 발표 주제에서 메인으로 다루는 것은 아니기 때문에 깊이 답변드리기는 어려운 점 참고 부탁드립니다!!
하기 답변은 [DAO vs Repository Patterns](https://www.baeldung.com/java-dao-vs-repository)을 참고하여 작성하였습니다.

DAO는 데이터 영속성을 추상화한 것으로, 데이터베이스에 가까운 하위 수준의 개념입니다. DAO는  data mapping/access 계층으로 사용되며, 쿼리를 숨길 수 있습니다.
Repository는 객체 집합을 추상화한 것으로, domain 객체에 더 가까운 상위 개념입니다. Repository는 domain과 data access 계층 사이에 있는 레이어이며, 데이터나 도메인 객체에 대한 정보를 숨길 수 있습니다. DAO는 리포지토리를 사용하여 구현할 수 없으나 Repository는 데이터베이스에 접근할 때 DAO를 사용할 수 있습니다.

**Reply by yeonkkk** [reply]
> Q3 : 모놀리식 애플리케이션을 자세히 알려주세요

모놀리식 애플리케이션은 다양한 구성 요소가 하나의 플랫폼에서 하나의 프로그램으로 결합된 소프트웨어 애플리케이션을 말합니다. 

* 모놀리식 애플리케이션의 장점
  * 개발이 간편하다.
  * 테스트가 간단하다.
  * 배포가 간단하다.
 
* 모놀리식 애플리케이션의 단점
  * 유지 및 관리가 어렵다.
  * 업데이트 시 전체 애플리케이션을 다시 시작해야한다.
  * 확장이 어려울 수 있다.
  * 안정성 문제

추가적으로 모놀리식 아키텍처가 궁금하시다면 아래 자료를 참고하시면 좋을 것 같습니다!
[Introduction to Monolithic Architecture and MicroServices Architecture](https://medium.com/koderlabs/introduction-to-monolithic-architecture-and-microservices-architecture-b211a5955c63)
